File: 11221/main.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <math.h>
#include <string.h>

void convert_line_to_alphas(char *src, char *dst){
    int i,j;
    for(i = j = 0; src[i]; i++)
        if(isalpha(src[i]))
            dst[j++] = tolower(src[i]);
    dst[j] = '\0';
}


void transpose_string(char *src, char *dst, int n){
    int i;
    for(i = 0; src[i]; i++)
        dst[i] = src[(i / n) + (i % n) * n];

    dst[i] = '\0';
}


int is_palindrome(char *string){
    int i, j;
    for(i = 0, j = strlen(string) - 1; i <= j; i++, j--)
        if((string[i] != string[j])) return 0;

    return 1;
}


int is_magic_square(char *alphas){

    int n, len = strlen(alphas);

    if(len == 0) return 0; /* 0*0 magic square */
    if((n = sqrt(len)) != sqrt(len)) return -1;

    char transpose[len + 1];
    transpose_string(alphas, transpose, n);

    return is_palindrome(alphas) && is_palindrome(transpose) ? n : -1;
}


void print(int index, int result){

    printf("Case #%d:\n", index + 1);

    if(result == -1) puts("No magic :(");
    else printf("%d\n", result);
}


int main()
{
    int N, i, result;
    char line[10001], alphas[10001];
    scanf("%d\n", &N);
    for(i = 0; i < N; i++){
        gets(line);
        convert_line_to_alphas(line, alphas);
        result = is_magic_square(alphas);
        print(i, result);
    }
    return 0;
}

File: 11221/Main.java

import java.util.Scanner;


class MagicSquarePalindrome{
    private String alphas;
    private String transpose;
    private int n;
    private int index;
    
    public MagicSquarePalindrome(String line, int index) {
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < line.length(); i++){
            if(Character.isAlphabetic(line.charAt(i)))
                sb.append(line.charAt(i));
        }
        alphas = sb.toString();
        this.index = index;
    }
    
    
    private boolean isPalindrome(String line){
        int i = 0, j = line.length() - 1;
        while(i <= j)
            if(line.charAt(i++) != line.charAt(j--))
                return false;
        return true;
    }
    
    private void transpose(){
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < alphas.length(); i++){
            sb.append(alphas.charAt((i / n) + (i % n) * n));
        }
        transpose = sb.toString();
    }

    private int isMagicSquare(){

        int len = alphas.length();

        if(len == 0) return 0;
        if((n = (int) Math.sqrt(len)) != Math.sqrt(len)) return -1;
        
        transpose();
        
        return isPalindrome(alphas) && isPalindrome(transpose) ? n : -1;
    }


    void print(){
        n = isMagicSquare();
        System.out.println("Case #" + ( index + 1) + ":");

        if(n == -1) System.out.println("No magic :(");
        else System.out.println(n);
    }


}

public class Main {
    public static void main(String[] args) {
        int N, i;
        Scanner sc = new Scanner(System.in);
        N = Integer.parseInt(sc.nextLine());
        
        for(i = 0; i < N; i++){
            MagicSquarePalindrome msp = new MagicSquarePalindrome(sc.nextLine(), i);
            msp.print();
        }
    }
}
